services/data/src/react/hooks/useDataQuery.ts

```diff
@@ -33,15 +33,18 @@ type QueryState = {
     refetchCallback?: (data: any) => void
 }
 
-export const useDataQuery = <TQueryResult = QueryResult>(
-    query: Query,
+export const useDataQuery = <
+    TQueryResultData extends QueryResultData<TQuery>,
+    TQuery extends Query<TQueryResultData> = Query<TQueryResultData>
```

Comment:
We could potentially remove the generic from `Query` to make this simpler. But that means you could specify a `Data` type and a query that does not match. 

The generic makes sure you cannot specify a query-key without typing it in the `Data`-type. It also errors if the `Data`-type contains a key not specified in the query. This errors gets quite long and could be confusing to consumers. So it may not be worth it to be that strict?